### DISCUSSION TITLE
How to access user_rois in the leafmap.Map class?

### DISCUSSION THREAD

**Discussion by tikitong** [opengeos/leafmap#152] [Q&A] [answered]
Hi @giswqs, 

I would like to access the geometries updated when drawing on the map, if I'm not mistaken they are updated with the attribute `self.user_rois` but how can I access them in the code after `m = leafmap.Map()`? 

https://github.com/giswqs/leafmap/blob/018311d1ab472fb3f710d75c8a153b2bf37a12ac/leafmap/leafmap.py#L104 

I would like for example to insert them in a code window that would be updated at each change with streamlit such as:
```
code = f"the drawn geometries are: {user_rois}"
st.code(code)
```
Thank you very much

**Comment by giswqs** [comment, marked as answer]
Unfortunately, `m.user_rois` won't work with streamlit. It is a one-way street. Once the the map is converted to html (`m_to_streamlit()`) and displayed in streamlit, all the bi-directional communication is gone. Whatever you do with the map in streamlit can't communicate back to python. The only workaround is to use the folium `draw_export` plugin to manually export the drawn geometries as a geojson and import it back to the app, just like what is implemented in https://github.com/giswqs/streamlit-geospatial.

**Reply by tikitong** [reply]
Thanks @giswqs! In addition I guess it is not possible to import a geojson file with `st.file_uploader` add it with `add_geojson` or `add_gdf` **and keep** the drawings made on the map. In the implemented examples https://github.com/giswqs/streamlit-geospatial the map is updated via m.to_streamlit() without taking into account the drawings which are there. 

It would be impossible to store the drawings made and add them to the geojson import, without having to export them first?

**Reply by giswqs** [reply]
This a limitation of the folium package. There is no way to retrieve the coordinates of the drawings programmatically. See https://github.com/python-visualization/folium/issues/1402. Using the `draw_export` plugin to export drawings manually is the only workaround that I am aware of. No mapping libraries support bidirectional communication in streamlit yet. The [streamlit-bokeh-events](https://github.com/ash2shukla/streamlit-bokeh-events) package supports some kinds of bidirectional communication with plots, but it doesn't work with maps yet.

**Reply by tikitong** [reply]
Well noted, thank you very much @giswqs

**Comment by zxdawn** [comment]
@giswqs Any updates after two years? It would be useful to pass user_rois bounds to search data vis pystac.

**Reply by giswqs** [reply]
Unfortunately, the answer is No for folium and streamlit. It is a one-way street. 
Try out leafmap with Solara. It supports bi-directional communication. You can enjoy all leafmap features with it using the ipyleaflet backend. 
https://huggingface.co/spaces/giswqs/solara-geospatial

**Reply by gomesdaciel** [reply]
@giswqs How do I retrieve the user_rois from the Map.element when using leafmap with Solara? I'd like to access the updated geometries when drawing on the map and assign them to a variable for executing a spatial join with GeoPandas.

**Reply by giswqs** [reply]
Look at this example: https://huggingface.co/spaces/giswqs/solara-maxar/blob/main/pages/01_morocco.py
